### DISCUSSION TITLE
Scheduled Tasks should be able to output to STDOUT - currently forced to /dev/null or file

### DISCUSSION THREAD

**Discussion by kingIZZZY** [laravel/framework#47448] [Ideas]
### Laravel Version

9.29.0

### PHP Version

8.1.20

### Description

Scheduled Tasks seem to [force output to /dev/null](https://github.com/laravel/framework/blob/585da0913e907fd54941260860ae3d7d4be8e8cb/src/Illuminate/Console/Scheduling/Event.php#L183) unless a local disk file is provided to write output to.

On Docker-based or similar deploy-code / spin-up-instances hosting services like [heroku.com](heroku.com) or [fly.io](fly.io) etc. local files are usually ephemeral and thus local log files are quite useless.
These hosts usually offer a live scrolling app output console log that can be monitored, very useful to catch live exceptions or view logged notices/warnings etc.

It would seem very desirable and necessary to have the option to let `schedule:run` allow laravel Schedule Events' output to proceed on to STDOUT along with all other app output. Suppressing it to /dev/null with no option to allow it through to STDOUT seems like blocking useful options for many deployment environments.

### Steps To Reproduce

- Create some Artisan commands or console kernel callback schedules
- Have these tasks/commands produce useful / informative / debugging output, whether laravel Log output, or artisan command `->line()` or `->table()` or whatnot
- Have these tasks/commands run as a Scheduled Task
- Never see your important output because the output has either been forced down the /dev/null chute, or into a local file that might not be there when you try to log in and read it

Some relevant links
- https://stackoverflow.com/questions/46586916/how-can-i-redirect-output-to-stdout-with-laravel-scheduler-command
- https://github.com/moby/moby/issues/19616#issuecomment-174355979

**Comment by kingIZZZY** [comment]
## Workaround

If your laravel deployment with docker is doing things the `supervisord` way like my project, here's how to get commands & scheduled tasks to show up in docker container output:

- Create a [named pipe (fifo)](https://en.wikipedia.org/wiki/Named_pipe) somewhere in the web directory, ex.  `mkfifo /path/to/laravel/storage/logs/stdout.fifo`
- `chown` this fifo it to the web user, so commands invoked from web or web-user cron can output to it
- Add a supervisord config to simply `cat /path/to/the/fifo`. It will ~literally continuously follow the fifo waiting for data and~ simply outputting to its stdout which goes to container stdout. Edit: OK no it will cat once when there's some data and then exit. So just put it in a 1-liner `while true; do cat /path/to/the/fifo; sleep 1; done` loop for good measure
- Have artisan commands output to the fifo. ex. `$schedule->command('dosomething:foo --option=1')->sendOutputTo(storage_path('logs/stdout.fifo'))`

**Reply by willrowe** [reply]
@kingIZZZY you could use `tail` instead of `cat`: `tail --follow --lines=0 --quiet /path/to/the/fifo`

**Comment by boiko-danylo** [comment]
Totally agree with the proposal. Current implementation breaks [12-factor](https://12factor.net/) methodology.

**Reply by henzeb** [reply]
While I think it's a good thing it's directed to a file or /dev/null by default, I too think we should be able to direct it to anywhere we want it too. Not just files or stdout...

**Comment by malsatin** [comment]
Previously that issue was mentioned in https://github.com/laravel/framework/discussions/38429
And discussion appeared in https://github.com/laravel/ideas/issues/126

Also, one of the solutions is to extend comand description like
```php
$schedule->command(TestLogCommand::class)->everyMinute()->thenWithOutput(function (Stringable $output) {
    echo $output;
});
```

**Reply by kingIZZZY** [reply]
I don't believe this solution works (I might even have tried it)
`echo` just outputs to the forced hardcoded `/dev/null` no?

Or maybe I misunderstand what you mean by "_solution to extend command description like ..._"

**Reply by malsatin** [reply]
> echo just outputs to the forced hardcoded /dev/null no?

No, it writes to something like `/dev/stdout`, that is default PHP behavior.

The reason, why current Laravel logger implementation writes logs either to `/dev/null` or to the specified file is not because Laravel overwrites standard PHP `echo` target from `/dev/stdout` to `dev/null`, but because Laravel don't try to `echo` the log line and they write the file directly to some file.

By adding `thenWithOutput` you force Laravel handler to provide you all log lines and you can then write them to stdout by yourself.
In our case it works as expected and solves the issue (although not in the most elegant way).

**Reply by kingIZZZY** [reply]
No I don't think that's correct... whatever you do with `thenWithOutput` ends up being redirected to either `/dev/null` or to the supplied file path, blocked from outputting to PHP standard output.

**Reply by malsatin** [reply]
Can you provide examples, which variants of `thenWithOutput` usage have you tried and to what results it led to?

We use the solution, provided in the comment in production and right now it works fine.
More particular example may be (this is a piece of code from my current project):
```php
protected function schedule(Schedule $schedule): void
{
    $showOutput = static function (Stringable $output): void {
        echo $output;
    };
    $schedule->command(RotateStorefrontsCommand::class)->everyTenMinutes()->thenWithOutput($showOutput);
}
```

**Comment by phpguru** [comment]
This is a challenge to understand when deploying to k8s - my `schedule:run` is on its own pod, but the commands its running don't log output because 
```
php artisan schedule:run
```
becomes
```
  2023-09-20 16:15:00 Running ['artisan' app:my-command] . 142ms DONE
  ⇂ '/usr/local/bin/php' 'artisan' app:my-command > '/dev/null' 2>&1
```
so Datadog and K8s have no way to get the command's logs.

**Comment by ghost** [comment]
Run in background scheduled task is not compatible with stdout logging in laravel. See https://github.com/laravel/framework/discussions/50085#discussioncomment-8495648

**Comment by p1ratrulezzz** [comment]
Instead of building a bicycle with macroses I've ended up using the patch for laravel/framework

<details><summary>scheduler_dont_truncate_output.patch</summary>
<p>

```php
--- a/src/Illuminate/Console/Scheduling/CommandBuilder.php
+++ b/src/Illuminate/Console/Scheduling/CommandBuilder.php
@@ -33,7 +33,7 @@
         $output = ProcessUtils::escapeArgument($event->output);

         return $this->ensureCorrectUser(
-            $event, $event->command.($event->shouldAppendOutput ? ' >> ' : ' > ').$output.' 2>&1'
+            $event, $event->command.' 2>&1'
         );
     }

@@ -56,7 +56,7 @@
         }

         return $this->ensureCorrectUser($event,
-            '('.$event->command.$redirect.$output.' 2>&1 ; '.$finished.' "$?") > '
+            '('.$event->command.' 2>&1 ; '.$finished.' "$?") > '
             .ProcessUtils::escapeArgument($event->getDefaultOutput()).' 2>&1 &'
         );
     }

--- a/src/Illuminate/Console/Scheduling/ScheduleRunCommand.php
+++ b/src/Illuminate/Console/Scheduling/ScheduleRunCommand.php
@@ -132,6 +132,8 @@
             }

             $this->eventsRan = true;
+
+            $this->handleEventOutput($event);
         }

         if ($events->contains->isRepeatable()) {
@@ -145,6 +147,18 @@
         }
     }

+    protected function handleEventOutput(Event $event): void
+    {
+        if (!empty($event->outputBufferStdout)) {
+            $this->components->info($event->outputBufferStdout);
+            $this->newLine();
+        }
+
+        if (!empty($event->outputBufferStderr)) {
+            $this->components->error($event->outputBufferStderr);
+        }
+    }
+
     /**
      * Run the given single server event.
      *
@@ -252,6 +266,8 @@
                 }

                 $this->eventsRan = true;
+
+                $this->handleEventOutput($event);
             }

             Sleep::usleep(100000);

--- a/src/Illuminate/Console/Scheduling/Event.php
+++ b/src/Illuminate/Console/Scheduling/Event.php
@@ -128,6 +128,16 @@
      */
     public $shouldAppendOutput = false;

+    /**
+     * @var string The output buffer for stdout.
+     */
+    public $outputBufferStdout = '';
+
+    /**
+     * @var string The output buffer for stderr.
+     */
+    public $outputBufferStderr = '';
+
     /**
      * The array of callbacks to be run before the event is started.
      *
@@ -289,7 +299,13 @@
     {
         return Process::fromShellCommandline(
             $this->buildCommand(), base_path(), null, null, null
-        )->run();
+        )->run(function($type, $buffer) {
+            if ($type === Process::ERR) {
+                $this->outputBufferStderr .= $buffer;
+            } else {
+                $this->outputBufferStdout .= $buffer;
+            }
+        });
     }

     /**

``` 

</p>
</details> 

To add patch automatically while composer install I've used the composer package *cweagans/composer-patches*

Here is my patch definition in 

<details><summary>composer.json</summary>
<p>

```json
  ...
 "extra": {
        "laravel": {
            "dont-discover": []
        },
        "patches": {
            "laravel/framework": {
                "scheduler_dont_truncate_output": "./patches/scheduler_dont_truncate_output.patch"
            }
        },
        "enable-patching": true
    },

 ...

``` 

</p>
</details> 

Maybe only stdout matter since it used 2>&1 directive. But I hope this might help someone.

The scheduler run command is really stupidly written. Logging to a file breaks 12factor app conseption. I hope in a newer version of the framework something good will be done with it.

**Reply by macropay-solutions** [reply]
@p1ratrulezzz we also had this issue and we adopted this solution https://github.com/laravel/framework/discussions/50085#discussioncomment-8495648 because even if we don't send them to /dev/null, the stdout is not catching anything.

Is your solution bringing something new (you managed to print to stdout from a bg proc_open process) or just the patch vs macros?

**Comment by abodnar** [comment]
Looks like whoever was working on Laravel Cloud realized that people might want to actually see the output from commands. So they made this change.

https://github.com/laravel/framework/blob/94c26967f8f7eebd61af49a90f1ee49b1bb6758b/src/Illuminate/Console/Scheduling/Event.php#L198-L207

Now I get that sometimes it takes encountering the situation yourself to realize that the way it is set up, doesn't work. But tying this to only for Laravel Cloud seems like a very narrow solution and really should be broadened to support those that are running Laravel in other environments.

**Reply by macropay-solutions** [reply]
Very interesting.
So Laravel is stepping on nextJs's footprints like Vercel. https://github.com/laravel/framework/discussions/55533

**Reply by davidharting** [reply]
Agreed that this is odd, but I'm assuming good intent here.

I hope to see this as a generic configuration option in the near future. Every containerized environment I've worked in wanted logs for everything to stdout to get picked up by a log ingestion process.

**Reply by macropay-solutions** [reply]
Update for laravel 9-12 https://medium.com/p/06d5db84f6e6

**Reply by phpguru** [reply]
> whoever was working on Laravel Cloud 

It was @taylorotwell in #53623 actually, and from the looks of it, simply setting LARAVEL_CLOUD=1 in your environment would start spewing logs from job runners to STDOUT, if I am reading that right. 

Now the question is, can we format all the output as JSON or are we still forced to get unstructured text intermixed in them?

**Reply by abodnar** [reply]
Yeah, I don't remember why I didn't write Taylor. Might have been cause I wasn't sure if it was him just merging the code or if he had implemented it.

It does seem like that's all you would need to do. I haven't had a chance to test it out.

I would imagine that if output wasn't coming through fully as JSON then there might be additional changes? 🤷

**Comment by thierry2015** [comment]
I just submitted https://github.com/laravel/framework/pull/56611, which doesn't solve all of the above, but might simplify how we control scheduled events' output

**Reply by macropay-solutions** [reply]
@thierry2015 you should mention in your PR that it will work only without runInBackground() for STDOUT.

**Reply by thierry2015** [reply]
Description updated, thanks. That being said, it would provide the same result as if one was to modify a single job output, which is why I haven't tested for these cases (out of scope)
